File: 207-course-schedule/course-schedule.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function(numCourses, prerequisites) {
    const indegree = new Array(numCourses).fill(0)

    for (const [course, prereq] of prerequisites) {
        indegree[course]++
    }

    const queue = []
    for (let i = 0; i < numCourses; i++) {
        if (indegree[i] === 0) {
            queue.push(i)
        }
    }

    let count = 0

    while (queue.length > 0) {
        const course = queue.shift()
        count++

        for (const [c, p] of prerequisites) {
            if (p === course) {
                indegree[c]--
                if (indegree[c] === 0) {
                    queue.push(c)
                }
            }
        }
    }
    return count === numCourses
};

File: 207-course-schedule/course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adj_list = [[] for _ in range(numCourses)]
        for course, prereq in prerequisites:
            adj_list[prereq].append(course)

        state = [0] * numCourses

        def dfs(course):
            if state[course] == 1:
                return False
            if state[course] == 2:
                return True

            state[course] = 1

            for neighbor in adj_list[course]:
                if not dfs(neighbor):
                    return False
            
            state[course] = 2
            return True

        for i in range(numCourses):
            if state[i] == 0:
                if not dfs(i):
                    return False

        return True